### ISSUE TITLE
[ASK] How to run with docker in  GPU environment

### ISSUE THREAD

**Issue by ljy0ustc** [microsoft/recommenders#1526]
I follow the start-up of docker in GPU environment like this:
![image](https://user-images.githubusercontent.com/87422961/133538392-ed0f44f8-ed5d-412e-9108-c3832b2d0c9d.png)
I succeed in running it. However when I check the GPU-util, it appears that it doesn't run on GPU.
![image](https://user-images.githubusercontent.com/87422961/133538554-df482745-fb47-4590-b828-02aabb1a3696.png)
How can I use my GPU?

**Comment by ljy0ustc** [comment]
The GPU-util mentioned above is when I run the examples/02_model_content_based_filtering/dkn_deep_dive.ipynb.

**Comment by gramhagen** [comment]
do you see any gpu available by running this code?
`tf.test.is_gpu_available()`

**Comment by ljy0ustc** [comment]
Yes.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "microsoft/recommenders"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年9月16日(星期四) 晚上9:28
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [microsoft/recommenders] [ASK] How to run with docker in  GPU environment (#1526)





 
do you see any gpu available by running this code?
 tf.test.is_gpu_available()
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by lhojun** [comment]
I'm in the same situation.

In my opinion CUDA cannot be used in RTX30xx
because these codes are based on tf1.x (RTX requires CUDA >= 11)
is this right?

if yes, is there any other way to run with GPU environment on the RTX 30xx?

**Comment by gramhagen** [comment]
You can try installing tensorflow>=2.5.0 and changing the code where tf is used
`import tensorflow as tf` -> `import tensorflow.compat.v1 as tf`

i have no idea if the v1 compat library covers all aspects of the code we have in the repo though

**Comment by lhojun** [comment]
I run the examples/00_quick_start/nrms_MIND.ipynb
it contain 'contrib' attribute
tensorflow guide says it was not helpful to use tensorflow.compat.v1 for 'contrib'

**Comment by gramhagen** [comment]
Ah,  I see.  Unfortunately some of this code needs to be ported to tf 2. The short term alternative is to use cuda 10.2. See conda gpu env instructions in SETUP.md

**Comment by anargyri** [comment]
I am not sure there is an issue, because you should be inside a docker container using CUDA 10, see https://github.com/microsoft/recommenders/blob/27709229cdc4aa7d39ab715789f093a2d21d2661/tools/docker/Dockerfile#L94
The output of nvidia-smi that you see looks ok, it doesn't matter that it says CUDA version 11.2. There will be no running process if you are not running any of the code that uses the GPU .
If `tf.test.is_gpu_available()` returns True, then you can run the notebooks.

You can double check by running the tests e.g. like this https://github.com/microsoft/recommenders/tree/main/tools/docker#running-tests-with-docker
```
docker run -it recommenders:gpu bash -c 'pip install pytest; \
pip install pytest-cov; \
apt-get install -y git; \
git clone https://github.com/microsoft/recommenders.git; \
cd recommenders; \
pytest tests/unit -m "not spark and gpu and not notebooks"'
```
You can also run the notebooks from Jupyter.

**Comment by whatactuallyis** [comment]
```
Dockerfile:118
--------------------
 116 |     SHELL ["/bin/bash", "--login", "-c"]
 117 |
 118 | >>> RUN if [ "${VIRTUAL_ENV}" = "conda" ] ; then conda install python=3.7; fi
 119 |     RUN if [ "${VIRTUAL_ENV}" = "venv" ] ; then apt-get -y install python3.7; \
 120 |         apt-get -y install python3-pip; \
--------------------
ERROR: failed to solve: process "/bin/bash --login -c if [ \"${VIRTUAL_ENV}\" = \"conda\" ] ; then conda install python=3.7; fi" did not complete successfully: exit code: 1
```

I couldn't follow why the docker file contains a python3.7 setup while the actual README suggests 3.9. Any help would be appreciated.

**Comment by whatactuallyis** [comment]
> In the meanwhile replacing 3.7 with 3.9 will likely work.

You're right. Thanks! Btw, I also wonder whether the system is also flexible for the CUDA version. For instance, I have 11.8 in my local. I am not an expert on CUDA. Therefore, not sure whether I can run different versions with a single GPU.

**Comment by anargyri** [comment]
I suspect it should be ok, although I don't think we have tested 11.8. 
The main dependence is TensorFlow, and Cuda 11.2 should work according to this
https://www.tensorflow.org/install/source#tested_build_configurations
